### PR TITLE
Add support for removing universes/entire chains via RemoveSecurity/UniverseManager.Remove

### PR DESCRIPTION
**Pull request by mchandschuh**
#### Description

In support of #1977 we need to ensure we're able to add/remove derivative universes properly. This change implements universe removal coupled with a regression algorithm that adds and removes an option chain universe from the algorithm via `OnData`.

#### Related Issue

See #1977 (next PR will actually resolve the issue)

#### Motivation and Context

Aiming to support universe selection models that can add/remove

#### Requires Documentation Change

Yup, will want to show in documentation that you can now remove an entire option chain and all child contracts via `RemoveSecurity(optionChainSymbol)`

#### How Has This Been Tested?

Extensively tested locally as well as an extremely thorough regression test with many checks within the algorithm related to expected securities/data

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed. (those that were passing :) )
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Util/LinqExtensions.cs [hunk 1]
```diff
@@ -315,5 +315,19 @@ public static IEnumerable<IEnumerable<T>> GroupAdjacentBy<T>(this IEnumerable<T>
...
+        /// This method uses sets to improve performance and also uses lazy evaluation so if a
+        /// difference is found, true is immediately returned and evaluation is halted.
+        /// </summary>
+        /// <typeparam name="T">The item type</typeparam>
+        /// <param name="left">The left set</param>
+        /// <param name="right">The right set</param>
+        /// <returns>True if there are any differences between the two sets, false otherwise</returns>
+        public static bool HasDiffs<T>(this ISet<T> left, ISet<T> right)
+        {
+            return left.Except(right).Any() || right.Except(left).Any();
```
Comment (StefanoRaggi) [inline Common/Util/LinqExtensions.cs]: 👍 Nice method, perhaps could be named `HasDifferences` or `IsDifferentFrom` instead ?

File: Common/SymbolRepresentation.cs [hunk 2]
```diff
@@ -180,13 +196,32 @@ public static string GenerateFutureTicker(string underlying, DateTime expiration
         /// <param name="right">Option right</param>
         /// <param name="strikePrice">Option strike</param>
         /// <param name="expiration">Option expiration date</param>
-        /// <returns></returns>
+        /// <returns>The OSI ticker representation</returns>
         public static string GenerateOptionTickerOSI(string underlying, OptionRight right, decimal strikePrice, DateTime expiration)
```
Comment (jaredbroad) [inline Common/SymbolRepresentation.cs]: 👍

File: Engine/DataFeeds/FileSystemDataFeed.cs [hunk 3]
```diff
@@ -250,14 +250,25 @@ public bool RemoveSubscription(SubscriptionDataConfig configuration)
 
             if (_subscriptions.TryGetValue(configuration, out subscription))
             {
+                // don't remove universe subscriptions immediately, instead mark them as disposed
+                // so we can turn the crank one more time to ensure we emit security changes properly
+                if (subscription.IsUniverseSelectionSubscription && subscription.Universe.DisposeRequested)
```
Comment (jaredbroad) [inline Engine/DataFeeds/FileSystemDataFeed.cs]: Interesting... Does the security data arrive for the last crank?

File: Common/Util/LinqExtensions.cs (hunk 1 above)
Comment (mchandschuh) [reply Common/Util/LinqExtensions.cs]: agreed, there's no need for us to be short with method names :)

File: Algorithm/QCAlgorithm.cs [hunk 4]
```diff
@@ -1597,21 +1597,57 @@ public Crypto AddCrypto(string ticker, Resolution resolution = Resolution.Minute
...
+                        if (!UniverseManager.Any(ukvp => !ReferenceEquals(ukvp.Value, universe) && ukvp.Value.Members.ContainsKey(underlying.Symbol)))
+                        {
+                            RemoveSecurity(underlying.Symbol);
+                        }
+                    }
 
+                    // remove child securities (option contracts for option chain universes)
+                    foreach (var child in universe.Members.Values)
+                    {
+                        RemoveSecurity(child.Symbol);
```
Comment (mchandschuh) [inline Algorithm/QCAlgorithm.cs]: Need to add a check before removal to ensure it's not in an other universe

File: Algorithm/QCAlgorithm.cs [hunk 5]
```diff
@@ -1597,21 +1597,57 @@ public Crypto AddCrypto(string ticker, Resolution resolution = Resolution.Minute
...
+                    }
+
+                    // finally, dispose and remove the canonical security from the universe manager
+                    UniverseManager.Remove(symbol);
```
[the remark this answers is not in the document]
Comment (mchandschuh) [reply Algorithm/QCAlgorithm.cs]: thanks! :)

**Pull request merged**